internal/recover/recover.go

```diff
@@ -57,43 +65,272 @@ func GetDqliteClusterMembers(filesystem *sys.OS) ([]cluster.DqliteMember, error)
+	err = daemonConfig.Load()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to read %q: %w", path, err)
+		return fmt.Errorf("Failed to load daemon.yaml: %w", err)
 	}
 
-	nodeInfo, err := store.Get(context.Background())
+	daemonConfig.SetAddress(newAddress)
+	err = daemonConfig.Write()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to read from node store: %w", err)
+		return fmt.Errorf("Failed to update daemon.yaml: %w", err)
 	}
 
-	return nodeInfo, nil
+	return nil
 }
 
 // ReadTrustStore parses the trust store. This is not thread safe!
-func ReadTrustStore(dir string) (*trust.Remotes, error) {
+func readTrustStore(dir string) (*trust.Remotes, error) {
 	remotes := &trust.Remotes{}
 	err := remotes.Load(dir)
 
 	return remotes, err
 }
 
+// Update the trust store with the new member addresses.
+func updateTrustStore(dir string, members []cluster.DqliteMember) error {
+	remotes, err := readTrustStore(dir)
+	if err != nil {
+		return err
+	}
+
+	remotesByName := remotes.RemotesByName()
+
+	trustMembers := make([]types.ClusterMember, 0, len(members))
+	for _, member := range members {
+		cert := remotesByName[member.Name].Certificate
+		addr, err := netip.ParseAddrPort(member.Address)
+		if err != nil {
+			return fmt.Errorf("Invalid address %q: %w", member.Address, err)
+		}
+
+		trustMembers = append(trustMembers, types.ClusterMember{
+			ClusterMemberLocal: types.ClusterMemberLocal{
+				Name:        member.Name,
+				Address:     types.AddrPort{AddrPort: addr},
+				Certificate: cert,
+			},
+		})
+	}
+
+	err = remotes.Replace(dir, trustMembers...)
+	if err != nil {
+		return fmt.Errorf("Update trust store at %q: %w", dir, err)
+	}
+
+	return nil
+}
+
 // ValidateMemberChanges compares two arrays of members to ensure:
 // - Their lengths are the same.
-// - Members with the same name also use the same ID and address.
+// - Members with the same name also use the same ID.
+// - There is at least one voter in newMembers.
+// - All the newMembers addresses can be parsed to a netip.AddrPort.
+// - There are no duplicate addresses.
 func ValidateMemberChanges(oldMembers []cluster.DqliteMember, newMembers []cluster.DqliteMember) error {
 	if len(newMembers) != len(oldMembers) {
 		return fmt.Errorf("members cannot be added or removed")
 	}
 
+	countVoters := 0
+	addrs := make(map[netip.AddrPort]bool)
+
 	for _, newMember := range newMembers {
 		memberValid := false
+
+		if newMember.Role == "voter" {
+			countVoters += 1
+		}
+
+		addr, err := netip.ParseAddrPort(newMember.Address)
+		if err != nil {
+			return fmt.Errorf("Invalid address %q: %w", newMember.Address, err)
+		}
+
+		if addrs[addr] {
+			return fmt.Errorf("Duplicate address %q", addr)
+		}
+
+		addrs[addr] = true
+
 		for _, oldMember := range oldMembers {
-			// FIXME: Allow changing member addresses as part of cluster recovery
 			membersMatch := newMember.DqliteID == oldMember.DqliteID &&
-				newMember.Name == oldMember.Name &&
```

Comment:
Small nit here, `memberValid` only exists to be assigned the value of `membersMatch` if it's true. 

You can simplify it a bit like this:

```go
memberValid := false
for _, oldMember := range oldMembers {
  memberValid = newMember.DqliteID == oldMember.DqliteID && newMember.Name == oldMember.Name
  if memberValid {
    break
  }
}
```